### DISCUSSION TITLE
Github Action builds with spark failing with java.net.BindException with spark commands

### DISCUSSION THREAD

**Discussion by nalinigans** [community/community#26585] [Actions]
<p>We started seeing this failure sporadically with our spark builds ubuntu 18.04, but it shows up all the time now. Fixed the <code>spark 3</code> tests by explicitly setting <code>SPARK_LOCAL_IP=127.0.0.1</code> in spark_env. But, the <code>spark 2</code> builds are still broken even with the env explicitly set and calling spark-submit with <code>--conf "spark.driver.bindAddress=127.0.0.1</code>. Running <code>lsof -iTCP -sTCP:LISTENe</code> before spark-submit show all the 4K ports to be free. See this <a href="https://github.com/GenomicsDB/GenomicsDB/actions/runs/764404954" rel="noopener nofollow ugc">broken build</a> for instance.</p>
<p>Jenkins builds seem to be fine.</p>
<pre><code class="lang-auto">2021-04-19T17:42:28.3311657Z 21/04/19 17:42:27 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
...
2021-04-19T17:42:28.3342144Z 21/04/19 17:42:27 ERROR SparkUI: Failed to bind SparkUI
2021-04-19T17:42:28.3344354Z java.net.BindException: Cannot assign requested address: Service 'SparkUI' failed after 16 retries (starting from 4040)! Consider explicitly setting the appropriate port for the service 'SparkUI' (for example spark.ui.port for SparkUI) to an available port or increasing spark.port.maxRetries.
</code></pre>
<p>Any pointers to  working around this issue for <code>spark 2</code>?</p>

**Comment by nalinigans** [comment]
<p>Also see</p><aside class="onebox githubpullrequest">
  <header class="source">
      <a href="https://github.com/apache/spark/pull/31983#issuecomment-811035931" target="_blank" rel="noopener nofollow ugc">github.com/apache/spark</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Pull Request">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 12 16" aria-hidden="true"><path d="M11 11.28V5c-.03-.78-.34-1.47-.94-2.06C9.46 2.35 8.78 2.03 8 2H7V0L4 3l3 3V4h1c.27.02.48.11.69.31.21.2.3.42.31.69v6.28A1.993 1.993 0 0 0 10 15a1.993 1.993 0 0 0 1-3.72zm-1 2.92c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zM4 3c0-1.11-.89-2-2-2a1.993 1.993 0 0 0-1 3.72v6.56A1.993 1.993 0 0 0 2 15a1.993 1.993 0 0 0 1-3.72V4.72c.59-.34 1-.98 1-1.72zm-.8 10c0 .66-.55 1.2-1.2 1.2-.65 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2zM2 4.2C1.34 4.2.8 3.65.8 3c0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2z"></path></svg>
  </div>

  <div class="github-info-container">

      <details class="onebox-details">
        <summary class="onebox-details-summary">
          <h4>
            <a href="https://github.com/apache/spark/pull/31983" target="_blank" rel="noopener nofollow ugc">[SPARK-34882][SQL] Replace if with filter clause in RewriteDistinctAggregates</a>
          </h4>
        </summary>
        <p class="onebox-details-body">&lt;!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
--&gt;

### What changes were proposed in this pull request?
&lt;!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
--&gt;
Replaced the `agg(if (('gid = 1)) 'cat1 else null)` pattern in `RewriteDistinctAggregates` with `agg('cat1) FILTER (WHERE 'gid = 1)`

### Why are the changes needed?
&lt;!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
--&gt;
For aggregate functions, that do not ignore NULL values (`First`, `Last` or `UDAF`s) the current approach can return wrong results.

In the added UT there are no nulls in the input `testData`. The query returned `Row(0, 1, 0, 51, 100)` before this PR.

### Does this PR introduce _any_ user-facing change?
&lt;!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
--&gt;
Bugfix

### How was this patch tested?
&lt;!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
--&gt;
UT
</p>
      </details>

    <div class="branches">
      <code>apache:master</code> ← <code>tanelk:SPARK-34882_distinct_agg_filter</code>
    </div>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2021-03-28" data-time="20:11:20" data-timezone="UTC">08:11PM - 28 Mar 21 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/tanelk" target="_blank" rel="noopener nofollow ugc">
          <img alt="tanelk" src="https://user-images.githubusercontent.com/35076948/181098579-49145a21-8a55-494a-98f7-0f31a1fdd386.png" class="onebox-avatar-inline" width="20" height="20">
          tanelk
        </a>
      </div>

      <div class="lines" title="7 commits changed 2 files with 49 additions and 27 deletions">
        <a href="https://github.com/apache/spark/pull/31983/files" target="_blank" rel="noopener nofollow ugc">
          <span class="added">+49</span>
          <span class="removed">-27</span>
        </a>
      </div>
    </div>

  </div>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
<aside class="onebox githubpullrequest">
  <header class="source">
      <a href="https://github.com/mlflow/mlflow/pull/4227" target="_blank" rel="noopener nofollow ugc">github.com/mlflow/mlflow</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Pull Request">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 12 16" aria-hidden="true"><path d="M11 11.28V5c-.03-.78-.34-1.47-.94-2.06C9.46 2.35 8.78 2.03 8 2H7V0L4 3l3 3V4h1c.27.02.48.11.69.31.21.2.3.42.31.69v6.28A1.993 1.993 0 0 0 10 15a1.993 1.993 0 0 0 1-3.72zm-1 2.92c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zM4 3c0-1.11-.89-2-2-2a1.993 1.993 0 0 0-1 3.72v6.56A1.993 1.993 0 0 0 2 15a1.993 1.993 0 0 0 1-3.72V4.72c.59-.34 1-.98 1-1.72zm-.8 10c0 .66-.55 1.2-1.2 1.2-.65 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2zM2 4.2C1.34 4.2.8 3.65.8 3c0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2z"></path></svg>
  </div>

  <div class="github-info-container">

      <details class="onebox-details">
        <summary class="onebox-details-summary">
          <h4>
            <a href="https://github.com/mlflow/mlflow/pull/4227" target="_blank" rel="noopener nofollow ugc">Fix flaky spark session creation</a>
          </h4>
        </summary>
        <p class="onebox-details-body">Signed-off-by: harupy &lt;17039389+harupy@users.noreply.github.com&gt;

## What changes are proposed in this pull request?

A tempoarary workaround for #4229

I'm investigating why this issue suddenly started happening.

## How is this patch tested?

(Details)

## Release Notes

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

&lt;!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
--&gt;
&lt;a name="release-note-category"&gt;&lt;/a&gt;
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes
</p>
      </details>

    <div class="branches">
      <code>mlflow:master</code> ← <code>harupy:fix-spark-address-issue</code>
    </div>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2021-04-06" data-time="06:34:30" data-timezone="UTC">06:34AM - 06 Apr 21 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/harupy" target="_blank" rel="noopener nofollow ugc">
          <img alt="harupy" src="https://user-images.githubusercontent.com/35076948/181098580-94a429bb-42d5-4766-99f0-40ae80bd987f.png" class="onebox-avatar-inline" width="20" height="20">
          harupy
        </a>
      </div>

      <div class="lines" title="7 commits changed 3 files with 32 additions and 10 deletions">
        <a href="https://github.com/mlflow/mlflow/pull/4227/files" target="_blank" rel="noopener nofollow ugc">
          <span class="added">+32</span>
          <span class="removed">-10</span>
        </a>
      </div>
    </div>

  </div>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by nalinigans** [comment, marked as answer]
<p>The solution was to cleanup /etc/hosts that had numerous aliases for 127.0.0.1 and to assign both spark master and local ip to 127.0.0.1 in spark_env.sh, setting the shell env was not sufficient.</p>
<p>Anyone know why these exist in /etc/hosts for ubuntu 18.04 images?</p>
<pre><code class="lang-auto">127.0.0.1 stratum.antpool.com
127.0.0.1 gratified.approvalbureau.com
127.0.0.1 inpervious.choppedgreen.com
127.0.0.1 nonretractile.choppedgreen.com
127.0.0.1 asia.cryptonight-hub.miningpoolhub.com
...
127.0.0.1 hub.miningpoolhub.com
</code></pre>

**Comment by jvalkeal** [comment]
<p>Most likely attempt from github to break mining attacks by messing up hosts table.</p>
